src/generate-types.ts

```diff
@@ -261,1117 +83,1396 @@ const statusCodes = [
   504
 ];
 
-function expandedWildcardCode(code: string): number[] {
-  if (code === 'default') {
-    return [];
-  }
-  const wildcard = new RegExp(code.replace(/X/g, '.'));
-  return statusCodes.filter(code => wildcard.test('' + code));
-}
+// list of files for which we have an action to generate it already
+// to prevent calling an action to generate it again and causing maybe some race conditions
+const generatedFiles: Set<string> = new Set();
 
-function generateResponseType(opt: Options, op: string, responses: oas.ResponsesObject) {
-  if (!responses) {
-    return assert.fail('missing responses');
-  }
-  const allSpecifiedStatuses: number[] = [];
-  Object.keys(responses)
-    .map(expandedWildcardCode)
-    .map(codes =>
-      codes.map(code => {
-        assert(allSpecifiedStatuses.indexOf(code) < 0, 'duplicate status code ' + code);
-        allSpecifiedStatuses.push(code);
-      })
+export function run(options: Options) {
+  // eslint-disable-next-line no-console
+  info(`generating ${options.targetFile} from ${options.sourceFile}`);
+  deprecated(
+    options.externalOpenApiImports.length > 0,
+    "'externalOpenApiImports' is deprecated. Consider using 'resolve' instead"
+  );
+  deprecated(
+    options.externalOpenApiSpecs,
+    "'externalOpenApiSpecs' is deprecated. Consider using 'resolve' instead"
+  );
+
+  const state = {
+    cwd: path.dirname(options.sourceFile),
+    imports: {} as Record<string, string>,
+    actions: [] as Array<() => Promise<void>>
+  };
+  const builtins = generateBuiltins();
+  const types = generateComponents(options);
+  const queryTypes = generateQueryTypes(options);
+
+  const externals = generateExternals(
+    Object.entries(state.imports).map(([importAs, importFile]) => ({
+      importAs,
+      importFile: resolveModule(options.targetFile, importFile)
```

Comment:
we resolve the paths that are given relative to cwd from resolve to paths relative to the import file to make things right.